cmd/server/pactasrv/analysis.go

```diff
@@ -0,0 +1,343 @@
+		mutations = append(mutations, db.SetAnalysisArtifactAdminDebugEnabled(*request.Body.AdminDebugEnabled))
+	}
+	if request.Body.SharedToPublic != nil {
+		mutations = append(mutations, db.SetAnalysisArtifactSharedToPublic(*request.Body.SharedToPublic))
+	}
+	err = s.DB.UpdateAnalysisArtifact(s.DB.NoTxn(ctx), id, mutations...)
+	if err != nil {
+		return nil, oapierr.Internal("failed to update analysis artifact", zap.Error(err))
+	}
+	return api.UpdateAnalysisArtifact204Response{}, nil
+}
+
+func (s *Server) checkAnalysisArtifactAuthorization(ctx context.Context, id pacta.AnalysisArtifactID) error {
+	actorOwnerID, err := s.getUserOwnerID(ctx)
+	if err != nil {
+		return err
+	}
+	// Extracted to a common variable so that we return the same response for not found and unauthorized.
+	notFoundErr := func(fields ...zapcore.Field) error {
+		fs := append(fields, zap.String("analysis_artifact_id", string(id)))
+		return oapierr.NotFound("analysis artifact not found", fs...)
+	}
+	aa, err := s.DB.AnalysisArtifact(s.DB.NoTxn(ctx), id)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return notFoundErr(zap.Error(err))
+		}
+		return oapierr.Internal("failed to look up analysis artifact", zap.String("analysis_artifact_id", string(id)), zap.Error(err))
+	}
+	a, err := s.DB.Analysis(s.DB.NoTxn(ctx), aa.AnalysisID)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return notFoundErr(zap.Error(err))
+		}
+		return oapierr.Internal("failed to look up analysis for analysis artifact",
+			zap.String("analysis_id", string(aa.AnalysisID)),
+			zap.Error(err))
+	}
+	if a.Owner.ID != actorOwnerID {
+		return notFoundErr(
+			zap.Error(fmt.Errorf("analysis artifact does not belong to user")),
+			zap.String("owner_id", string(a.Owner.ID)),
+			zap.String("actor_id", string(actorOwnerID)),
+			zap.String("analysis_id", string(aa.AnalysisID)))
+	}
+	return nil
+}
+
+// Requests an anslysis be run
+// (POST /run-analysis)
+func (s *Server) RunAnalysis(ctx context.Context, request api.RunAnalysisRequestObject) (api.RunAnalysisResponseObject, error) {
+	ownerID, err := s.getUserOwnerID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	analysisType, err := conv.AnalysisTypeFromOAPI(&request.Body.AnalysisType)
+	if err != nil {
+		return nil, err
+	}
+
+	found := 0
+	var iID pacta.InitiativeID
+	var pgID pacta.PortfolioGroupID
+	var pID pacta.PortfolioID
+	if request.Body.InitiativeId != nil {
+		iID = pacta.InitiativeID(*request.Body.InitiativeId)
+		found++
+	}
+	if request.Body.PortfolioGroupId != nil {
+		pgID = pacta.PortfolioGroupID(*request.Body.PortfolioGroupId)
+		found++
+	}
+	if request.Body.PortfolioId != nil {
+		pID = pacta.PortfolioID(*request.Body.PortfolioId)
+		found++
+	}
+	if found == 0 {
+		return nil, oapierr.BadRequest("one of initiative_id, portfolio_group_id, or portfolio_id is required")
+	}
+	if found > 1 {
+		return nil, oapierr.BadRequest("only one of initiative_id, portfolio_group_id, or portfolio_id may be set")
+	}
```

Comment:
Note that these error messages are for us, they don't go to end-users. The end user just gets 400 `Bad Request`

Reply:
Great point. I'll file a bug to go through our BadRequest creations and add human-readable error message codes.

Reply:
#98